### ISSUE TITLE
Compile problem

### ISSUE THREAD

**Issue by felix1024** [gilphilbert/falk-pre-esp32#3]
Hi
i have recently found your modular preamp project at diyaudio.com. Now I wanted to compile the software for the advanced controller by myself. Until now I have no experience with Visual Studio Code and PlatformIO, just working with Arduino IDE.
So I setup VS Code, PlatformIO and ESP-IDF, cloned your code from GitHub and tried to build.
It happens:
Processing esp32dev (platform: espressif32; board: esp32dev; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via -v, --verbose option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
- framework-arduinoespressif32 @ 3.20011.230801 (2.0.11) 
- tool-esptoolpy @ 1.40501.0 (4.5.1) 
- toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 39 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ArduinoJson @ 6.21.3
|-- ESP Async WebServer @ 1.2.3+sha.f71e3d4
|-- U8g2 @ 2.35.7
|-- SPI @ 2.0.0
|-- Wire @ 2.0.0
|-- Preferences @ 2.0.0
|-- ESP32Encoder @ 0.10.1
|-- ESPmDNS @ 2.0.0
|-- SPIFFS @ 2.0.0
|-- Update @ 2.0.0
|-- WiFi @ 2.0.0
Building in release mode
Compiling .pio/build/esp32dev/src/display-ssd1322.cpp.o
Compiling .pio/build/esp32dev/src/falk-pre-conf.cpp.o
Compiling .pio/build/esp32dev/src/main.cpp.o
Compiling .pio/build/esp32dev/src/relay-input.cpp.o
Compiling .pio/build/esp32dev/src/relay-volume.cpp.o
Compiling .pio/build/esp32dev/src/rubik.c.o
Compiling .pio/build/esp32dev/src/wifi.cpp.o
Building .pio/build/esp32dev/bootloader.bin
Generating partitions .pio/build/esp32dev/partitions.bin
esptool.py v4.5.1
Creating esp32 image...
Merged 1 ELF section
Successfully created esp32 image.
Compiling .pio/build/esp32dev/libe68/AsyncTCP/AsyncTCP.cpp.o
Compiling .pio/build/esp32dev/lib1b6/FS/FS.cpp.o
Compiling .pio/build/esp32dev/lib1b6/FS/vfs_api.cpp.o
src/wifi.cpp: In member function 'void WiFiManager::loop()':
src/wifi.cpp:29:16: error: 'WiFi' was not declared in this scope
int status = WiFi.status();
^~~~
src/wifi.cpp:32:33: error: 'WL_CONNECT_FAILED' was not declared in this scope
if (disconnected && status == WL_CONNECT_FAILED) {
... lot of undeclared in scope ....

Maybe you could give me a hint prior to digging myself :) 
Appreciate your input
Kindly
Felix

PS: Great project!

**Comment by gilphilbert** [comment]
So `WL_CONNECT_FAILED` is a part of the underlying framework, if it's not being found then there's something wrong with the toolchain. I've just rebuilt my main machine, let me pull the code for this project again and see if I can replicate the issue. I only just updated this a couple of months back with a much better WiFi implementation, so I'm surprised by this!

**EDIT**: I completely missed this in your reply:
```
src/wifi.cpp:29:16: error: 'WiFi' was not declared in this scope
```

This means that PlatformIO isn't finding the Arduino headers, since that's a part of the Arduino framework for ESP32 devices.

I just set everything up from scratch, so this is what I did. Assuming you've got a "blank slate" machine, the setup should look like this:
- Install Visual Studio Code
- From Extensions, install **PlatformIO**
- If you're on Linux (Debian) and PlatformIO complains about not finding Python: `sudo apt install python3-venv`
- Wherever makes sense on your file system, run `git clone https://github.com/gilphilbert/falk-pa01-advanced.git`
- Open that folder in VS Code (`File` > `Open Folder`)
- In the terminal at the bottom, PlatformIO should be loading the libraries needed to compile the project. This will take a minute or two depending on your machine and internet connection
- Build the project (CTRL+SHIFT+B) or PlatformIO tab and select `Build` under `esp32dev` > `General`

Result should be:
```
Processing esp32dev (platform: espressif32; board: esp32dev; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20011.230801 (2.0.11) 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 39 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ArduinoJson @ 6.21.2
|-- ESP Async WebServer @ 1.2.3+sha.f71e3d4
|-- U8g2 @ 2.35.3
|-- SPI @ 2.0.0
|-- Wire @ 2.0.0
|-- Preferences @ 2.0.0
|-- ESP32Encoder @ 0.10.1
|-- ESPmDNS @ 2.0.0
|-- SPIFFS @ 2.0.0
|-- Update @ 2.0.0
|-- WiFi @ 2.0.0
Building in release mode
Compiling .pio/build/esp32dev/src/display-ssd1322.cpp.o
Compiling .pio/build/esp32dev/src/falk-pre-conf.cpp.o
Compiling .pio/build/esp32dev/src/main.cpp.o
Compiling .pio/build/esp32dev/src/relay-input.cpp.o
Compiling .pio/build/esp32dev/src/relay-volume.cpp.o
Compiling .pio/build/esp32dev/src/rubik.c.o
Compiling .pio/build/esp32dev/src/wifi.cpp.o
Building .pio/build/esp32dev/bootloader.bin
Generating partitions .pio/build/esp32dev/partitions.bin
Compiling .pio/build/esp32dev/lib391/AsyncTCP/AsyncTCP.cpp.o
esptool.py v4.5.1
Creating esp32 image...
Merged 1 ELF section
Successfully created esp32 image.
Compiling .pio/build/esp32dev/lib154/FS/FS.cpp.o
---------
Compiling .pio/build/esp32dev/FrameworkArduino/wiring_shift.c.o
Archiving .pio/build/esp32dev/libFrameworkArduino.a
Indexing .pio/build/esp32dev/libFrameworkArduino.a
Linking .pio/build/esp32dev/firmware.elf
Retrieving maximum program size .pio/build/esp32dev/firmware.elf
Checking size .pio/build/esp32dev/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]  14.8% (used 48364 bytes from 327680 bytes)
Flash: [========= ]  89.3% (used 936117 bytes from 1048576 bytes)
Building .pio/build/esp32dev/firmware.bin
esptool.py v4.5.1
Creating esp32 image...
Merged 25 ELF sections
Successfully created esp32 image.
=========================================================================================================================== [SUCCESS] Took 10.72 seconds ===========================================================================================================================
```

To flash the code to an ESP32, you'll first need to flash the firmware, then you'll need to generate and flash the SPIFFS image.

I will mention that without the screen and buttons connected the ESP32 will crash on boot, from memory (I haven't tried this in a while)

PlatformIO can be a little overwhelming at first if you're used to Arduino IDE. PlatformIO is far more sophisticated than Arduino IDE so if you have questions let me know.

**Comment by felix1024** [comment]
Hi again
I restarted from scratch removed and reinstalled step by step as you described above. Unfortunately with the same result:
 Executing task: platformio run --environment esp32dev 

Processing esp32dev (platform: espressif32; board: esp32dev; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20011.230801 (2.0.11) 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 39 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ArduinoJson @ 6.21.3
|-- ESP Async WebServer @ 1.2.3+sha.f71e3d4
|-- U8g2 @ 2.35.7
|-- SPI @ 2.0.0
|-- Wire @ 2.0.0
|-- Preferences @ 2.0.0
|-- ESP32Encoder @ 0.10.1
|-- ESPmDNS @ 2.0.0
|-- SPIFFS @ 2.0.0
|-- Update @ 2.0.0
|-- WiFi @ 2.0.0
Building in release mode
Compiling .pio/build/esp32dev/src/display-ssd1322.cpp.o
Compiling .pio/build/esp32dev/src/falk-pre-conf.cpp.o
Compiling .pio/build/esp32dev/src/main.cpp.o
Compiling .pio/build/esp32dev/src/relay-input.cpp.o
Compiling .pio/build/esp32dev/src/relay-volume.cpp.o
Compiling .pio/build/esp32dev/src/rubik.c.o
Compiling .pio/build/esp32dev/src/wifi.cpp.o
Building .pio/build/esp32dev/bootloader.bin
esptool.py v4.5.1
Creating esp32 image...
Merged 1 ELF section
Successfully created esp32 image.
Generating partitions .pio/build/esp32dev/partitions.bin
Compiling .pio/build/esp32dev/libe68/AsyncTCP/AsyncTCP.cpp.o
Compiling .pio/build/esp32dev/lib1b6/FS/FS.cpp.o
Compiling .pio/build/esp32dev/lib1b6/FS/vfs_api.cpp.o
src/wifi.cpp: In member function 'void WiFiManager::loop()':
src/wifi.cpp:29:16: error: 'WiFi' was not declared in this scope
   int status = WiFi.status();
                ^~~~
src/wifi.cpp:32:33: error: 'WL_CONNECT_FAILED' was not declared in this scope
   if (disconnected && status == WL_CONNECT_FAILED) {
                                 ^~~~~~~~~~~~~~~~~
src/wifi.cpp:32:33: note: suggested alternative: 'WS_CONNECTED'
   if (disconnected && status == WL_CONNECT_FAILED) {
                                 ^~~~~~~~~~~~~~~~~
                                 WS_CONNECTED
src/wifi.cpp: At global scope:
src/wifi.cpp:54:27: error: variable or field 'WiFiStationConnected' declared void
 void WiFiStationConnected(WiFiEvent_t event, WiFiEventInfo_t info){
                           ^~~~~~~~~~~
src/wifi.cpp:54:27: error: 'WiFiEvent_t' was not declared in this scope
src/wifi.cpp:54:27: note: suggested alternative: 'ip_event_t'
 void WiFiStationConnected(WiFiEvent_t event, WiFiEventInfo_t info){
                           ^~~~~~~~~~~
                           ip_event_t
src/wifi.cpp:54:46: error: 'WiFiEventInfo_t' was not declared in this scope
 void WiFiStationConnected(WiFiEvent_t event, WiFiEventInfo_t info){
                                              ^~~~~~~~~~~~~~~
src/wifi.cpp:58:16: error: variable or field 'WiFiGotIP' declared void
 void WiFiGotIP(WiFiEvent_t event, WiFiEventInfo_t info){
                ^~~~~~~~~~~
src/wifi.cpp:58:16: error: 'WiFiEvent_t' was not declared in this scope
src/wifi.cpp:58:16: note: suggested alternative: 'ip_event_t'
 void WiFiGotIP(WiFiEvent_t event, WiFiEventInfo_t info){
                ^~~~~~~~~~~
                ip_event_t
src/wifi.cpp:58:35: error: 'WiFiEventInfo_t' was not declared in this scope
 void WiFiGotIP(WiFiEvent_t event, WiFiEventInfo_t info){
                                   ^~~~~~~~~~~~~~~
src/wifi.cpp:88:30: error: variable or field 'WiFiStationDisconnected' declared void
 void WiFiStationDisconnected(WiFiEvent_t event, WiFiEventInfo_t info){
                              ^~~~~~~~~~~
src/wifi.cpp:88:30: error: 'WiFiEvent_t' was not declared in this scope
src/wifi.cpp:88:30: note: suggested alternative: 'ip_event_t'
 void WiFiStationDisconnected(WiFiEvent_t event, WiFiEventInfo_t info){
                              ^~~~~~~~~~~
                              ip_event_t
src/wifi.cpp:88:49: error: 'WiFiEventInfo_t' was not declared in this scope
 void WiFiStationDisconnected(WiFiEvent_t event, WiFiEventInfo_t info){
                                                 ^~~~~~~~~~~~~~~
src/wifi.cpp:109:15: error: variable or field 'ScanDone' declared void
 void ScanDone(WiFiEvent_t event, WiFiEventInfo_t info){
               ^~~~~~~~~~~
src/wifi.cpp:109:15: error: 'WiFiEvent_t' was not declared in this scope
src/wifi.cpp:109:15: note: suggested alternative: 'ip_event_t'
 void ScanDone(WiFiEvent_t event, WiFiEventInfo_t info){
               ^~~~~~~~~~~
               ip_event_t
src/wifi.cpp:109:34: error: 'WiFiEventInfo_t' was not declared in this scope
 void ScanDone(WiFiEvent_t event, WiFiEventInfo_t info){
                                  ^~~~~~~~~~~~~~~
src/wifi.cpp: In member function 'void WiFiManager::begin()':
src/wifi.cpp:114:3: error: 'WiFi' was not declared in this scope
   WiFi.setHostname(sysSettings.hostname.c_str());
   ^~~~
src/wifi.cpp:117:16: error: 'WiFiStationConnected' was not declared in this scope
   WiFi.onEvent(WiFiStationConnected, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_CONNECTED);
                ^~~~~~~~~~~~~~~~~~~~
src/wifi.cpp:117:38: error: 'WiFiEvent_t' has not been declared
   WiFi.onEvent(WiFiStationConnected, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_CONNECTED);
                                      ^~~~~~~~~~~
src/wifi.cpp:118:16: error: 'WiFiGotIP' was not declared in this scope
   WiFi.onEvent(WiFiGotIP, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_GOT_IP);
                ^~~~~~~~~
src/wifi.cpp:118:27: error: 'WiFiEvent_t' has not been declared
   WiFi.onEvent(WiFiGotIP, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_GOT_IP);
                           ^~~~~~~~~~~
src/wifi.cpp:119:16: error: 'WiFiStationDisconnected' was not declared in this scope
   WiFi.onEvent(WiFiStationDisconnected, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_DISCONNECTED);
                ^~~~~~~~~~~~~~~~~~~~~~~
src/wifi.cpp:119:41: error: 'WiFiEvent_t' has not been declared
   WiFi.onEvent(WiFiStationDisconnected, WiFiEvent_t::ARDUINO_EVENT_WIFI_STA_DISCONNECTED);
                                         ^~~~~~~~~~~
src/wifi.cpp:120:16: error: 'ScanDone' was not declared in this scope
   WiFi.onEvent(ScanDone, WiFiEvent_t::ARDUINO_EVENT_WIFI_SCAN_DONE);
                ^~~~~~~~
src/wifi.cpp:120:16: note: suggested alternative: 'scanning'
   WiFi.onEvent(ScanDone, WiFiEvent_t::ARDUINO_EVENT_WIFI_SCAN_DONE);
                ^~~~~~~~
                scanning
src/wifi.cpp:120:26: error: 'WiFiEvent_t' has not been declared
   WiFi.onEvent(ScanDone, WiFiEvent_t::ARDUINO_EVENT_WIFI_SCAN_DONE);
                          ^~~~~~~~~~~
src/wifi.cpp:130:13: error: 'WIFI_STA' was not declared in this scope
   WiFi.mode(WIFI_STA);
             ^~~~~~~~
src/wifi.cpp:130:13: note: suggested alternative: 'WIFI_IF_STA'
   WiFi.mode(WIFI_STA);
             ^~~~~~~~
             WIFI_IF_STA
src/wifi.cpp: In member function 'bool WiFiManager::enableAP()':
src/wifi.cpp:136:3: error: 'WiFi' was not declared in this scope
   WiFi.mode(WIFI_AP_STA);
   ^~~~
src/wifi.cpp:136:13: error: 'WIFI_AP_STA' was not declared in this scope
   WiFi.mode(WIFI_AP_STA);
             ^~~~~~~~~~~
src/wifi.cpp:136:13: note: suggested alternative: 'WIFI_IF_STA'
   WiFi.mode(WIFI_AP_STA);
             ^~~~~~~~~~~
             WIFI_IF_STA
src/wifi.cpp: In function 'void getNetworks()':
src/wifi.cpp:177:26: error: 'WiFi' was not declared in this scope
   int numberOfNetworks = WiFi.scanComplete();
                          ^~~~
Compiling .pio/build/esp32dev/lib8e4/WiFi/WiFi.cpp.o
src/wifi.cpp: In lambda function:
src/wifi.cpp:484:5: error: 'WiFi' was not declared in this scope
     WiFi.scanNetworks();
     ^~~~
src/wifi.cpp: In lambda function:
src/wifi.cpp:505:5: error: 'WiFi' was not declared in this scope
     WiFi.begin(temp_ssid.c_str(), temp_key.c_str());
     ^~~~
Compiling .pio/build/esp32dev/lib8e4/WiFi/WiFiAP.cpp.o
*** [.pio/build/esp32dev/src/wifi.cpp.o] Error 1
============================================================= [FAILED] Took 12.72 seconds =============================================================


If I'm going to check for problems I get still (src/wifi.cpp)
'WiFi' was not declared in this scope (ln 29, Col16)

Right click on WiFi Goto Declaration shows "no declaration for 'WiFi'"

I'm running VSCode 1.82.2 (Universal) on Mac OSX 10.15.7
python3 --version
Python 3.11.5

please help

**Comment by gilphilbert** [comment]
I don't have a Mac, so it's hard for me to work out exactly what's going on, but it's clear that your environment isn't able to find the `WiFi.h` header of the Arduino ESP32 library. It should be in your environment somewhere - by default, PlatformIO will install it in your home directory, on my machine that's:
`/home/user/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFi.h`

You should be able to find it with `find ~/ -name WiFi.h`. That'll let you know it's installed somewhere.

[I found someone else with a similar problem](/home/phill/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFi.h), they fixed it with the following build flag in `platformio.ini` (in the root of the project):
`lib_ldf_mode = deep`
I can't guarantee that'll fix it, but it's worth a shot. Other options for that flag are:
- `deep+`
- `chain` (this is the default)
- `chain+`

[This is in the PlatformIO documentation](https://docs.platformio.org/en/stable/librarymanager/ldf.html#ldf-mode)

**Comment by felix1024** [comment]
WiFi.h exists on path /User/felix/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFi.h

but I was not successful even with changing options.

So I have installed a virtual machine running Debian 12 on my Mac, installed VSCode and PlatformIO there and surprise everything worked as expected:
....
Linking .pio/build/esp32dev/firmware.elf
Retrieving maximum program size .pio/build/esp32dev/firmware.elf
Checking size .pio/build/esp32dev/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]  14.8% (used 48364 bytes from 327680 bytes)
Flash: [========= ]  89.3% (used 936117 bytes from 1048576 bytes)
Building .pio/build/esp32dev/firmware.bin
esptool.py v4.5.1
Creating esp32 image...
Merged 25 ELF sections
Successfully created esp32 image.
=============================================== [SUCCESS] Took 258.17 seconds ===============================================

The problem lies on Mac OS (probably some security features preventing access to files) if I find a solution I will let you know.
Thanks for your support
Kindly
Felix

**Comment by felix1024** [comment]
The issue is solved on MAC, the problem is file naming of wifi.h and wifi.cpp. The compiler cannot find WiFi.h because the  filesystem on MAC and Windows are **case insensitive**! 
So renaming wifi.cpp to wifimanager.cpp and wifi.h to wifimanager.h and changing includes. will fix the problem.

**Comment by gilphilbert** [comment]
Thanks for identifying this!

Fixed in commit 6cdd8e5

**Comment by felix1024** [comment]
hi gilphilbert,
i have studied your software, and your very nice design behind. I learned a lot and could benefit and improve my skills. (I'm not a professional programmer) I did this because i wanted to include some additions and changes. I ended up with partially rewritten all the webcode, javascript, html and css. I also designed new audio and controller hardware to serve all my requirements.

Additions/changes:
- using a ESP32 Devkit board (prevent smt soldering) with external WiFi Antenna
- now with 2 displays
- SPI serial RAM, battery buffered to save settings (23LCV512)
- IR receiver + software (using a nice Apple remote)
- Relay controlled analog tone control, loudness control and channel separated volume control
- VU-meters in software.
- replaced REST API by a websocket approach (all functions controlled local and remote are synchronized)
- Replaced SPIFF by LittleFS
- no differential audio (don't see any advantage)
- 6 audio modular inputs (4 line inputs adjustable, 1 MC/MM phono input) 1 spare for later digital input (DAC/BT)
- minimized cabeling

As give back i would like to provide the code to you (if you are interested) and asking if you don't mind if i publish it later on github? I'm still not sure if i'm going to publish the whole project in diyaudio.com like you did. Some comments/critics there are...

If you could provide me your email to felix@thommen.net i can send you more.

Kindly
Felix
